File: bindings/dart/lib/src/log/log.dart
/// Log configuration and copied native log records.
library;

/// Log severity emitted by MapLibre Native.
final class LogSeverity {
  const LogSeverity._(this.rawValue, this.name);

  /// Informational log record.
  static const info = LogSeverity._(1, 'info');

  /// Warning log record.
  static const warning = LogSeverity._(2, 'warning');

  /// Error log record.
  static const error = LogSeverity._(3, 'error');

  /// Creates a log severity from a raw native value.
  factory LogSeverity.fromRawValue(int rawValue) => switch (rawValue) {
    1 => info,
    2 => warning,
    3 => error,
    _ => LogSeverity._(rawValue, 'unknown($rawValue)'),
  };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Log event category emitted by MapLibre Native.
final class LogEvent {
  const LogEvent._(this.rawValue, this.name);

  /// General category.
  static const general = LogEvent._(0, 'general');

  /// Setup category.
  static const setup = LogEvent._(1, 'setup');

  /// Shader category.
  static const shader = LogEvent._(2, 'shader');

  /// Parse-style category.
  static const parseStyle = LogEvent._(3, 'parseStyle');

  /// Parse-tile category.
  static const parseTile = LogEvent._(4, 'parseTile');

  /// Render category.
  static const render = LogEvent._(5, 'render');

  /// Style category.
  static const style = LogEvent._(6, 'style');

  /// Database category.
  static const database = LogEvent._(7, 'database');

  /// HTTP request category.
  static const httpRequest = LogEvent._(8, 'httpRequest');

  /// Sprite category.
  static const sprite = LogEvent._(9, 'sprite');

  /// Image category.
  static const image = LogEvent._(10, 'image');

  /// OpenGL category.
  static const openGl = LogEvent._(11, 'openGl');

  /// JNI category.
  static const jni = LogEvent._(12, 'jni');

  /// Android category.
  static const android = LogEvent._(13, 'android');

  /// Crash category.
  static const crash = LogEvent._(14, 'crash');

  /// Glyph category.
  static const glyph = LogEvent._(15, 'glyph');

  /// Timing category.
  static const timing = LogEvent._(16, 'timing');

  /// Creates a log event from a raw native value.
  factory LogEvent.fromRawValue(int rawValue) => switch (rawValue) {
    0 => general,
    1 => setup,
    2 => shader,
    3 => parseStyle,
    4 => parseTile,
    5 => render,
    6 => style,
    7 => database,
    8 => httpRequest,
    9 => sprite,
    10 => image,
    11 => openGl,
    12 => jni,
    13 => android,
    14 => crash,
    15 => glyph,
    16 => timing,
    _ => LogEvent._(rawValue, 'unknown($rawValue)'),
  };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Copied native log record.
final class LogRecord {
  /// Creates a copied native log record.
  const LogRecord({
    required this.severity,
    required this.event,
    required this.code,
    required this.message,
  });

  /// Record severity.
  final LogSeverity severity;

  /// Event category.
  final LogEvent event;

  /// Native log code.
  final int code;

  /// Copied message.
  final String message;
}

/// Log callback run asynchronously on Dart's owner isolate.
typedef LogCallback = void Function(LogRecord record);

/// Log severity mask for asynchronous native log dispatch.
final class LogSeverityMask {
  /// Creates a log severity mask from raw C bits.
  const LogSeverityMask(this.bits);

  /// Info severity bit.
  static const info = LogSeverityMask(1 << 1);

  /// Warning severity bit.
  static const warning = LogSeverityMask(1 << 2);

  /// Error severity bit.
  static const error = LogSeverityMask(1 << 3);

  /// Native default mask: info and warning may be asynchronous.
  static const defaultMask = LogSeverityMask((1 << 1) | (1 << 2));

  /// All known severity bits.
  static const all = LogSeverityMask((1 << 1) | (1 << 2) | (1 << 3));

  /// Raw native mask bits.
  final int bits;

  /// Returns true when all [severity] bits are present in this mask.
  bool contains(LogSeverityMask severity) =>
      (bits & severity.bits) == severity.bits;
}

File: bindings/python/python/maplibre_native_ffi/log.py
"""Process-global logging callback APIs."""

from __future__ import annotations

from ._enum import UnknownIntEnum
from dataclasses import dataclass
from enum import IntFlag
from typing import Any

from . import _native

_LOG_RECEIVER_CREATE_KEY = object()


class LogSeverity(UnknownIntEnum):
    """Log severity values emitted by MapLibre Native."""

    INFO = 1
    WARNING = 2
    ERROR = 3


class LogSeverityMask(IntFlag):
    """Mask of severities MapLibre Native may dispatch asynchronously."""

    INFO = 1 << LogSeverity.INFO
    WARNING = 1 << LogSeverity.WARNING
    ERROR = 1 << LogSeverity.ERROR
    DEFAULT = INFO | WARNING
    ALL = INFO | WARNING | ERROR


class LogEvent(UnknownIntEnum):
    """Log event categories emitted by MapLibre Native."""

    GENERAL = 0
    SETUP = 1
    SHADER = 2
    PARSE_STYLE = 3
    PARSE_TILE = 4
    RENDER = 5
    STYLE = 6
    DATABASE = 7
    HTTP_REQUEST = 8
    SPRITE = 9
    IMAGE = 10
    OPENGL = 11
    JNI = 12
    ANDROID = 13
    CRASH = 14
    GLYPH = 15
    TIMING = 16


@dataclass(frozen=True, slots=True)
class LogRecord:
    """Copied MapLibre Native log record."""

    severity: LogSeverity
    event: LogEvent
    code: int
    message: str

    @classmethod
    def _from_native(cls, raw: dict[str, Any]) -> "LogRecord":
        """Build a copied log record from private native values."""
        return cls(
            severity=LogSeverity(raw["severity"]),
            event=LogEvent(raw["event"]),
            code=raw["code"],
            message=raw["message"],
        )


class LogReceiver:
    """Bounded queue receiving copied process-global native log records."""

    def __init__(self, native: Any, *, _create_key: object | None = None) -> None:
        if _create_key is not _LOG_RECEIVER_CREATE_KEY:
            msg = "LogReceiver instances are created by set_log_callback"
            raise TypeError(msg)
        self._native = native

    @classmethod
    def _from_native(cls, native: Any) -> "LogReceiver":
        return cls(native, _create_key=_LOG_RECEIVER_CREATE_KEY)

    @property
    def dropped_record_count(self) -> int:
        """Return records dropped because the queue was full."""
        return int(self._native.dropped_record_count)

    def poll_record(self) -> LogRecord | None:
        """Return one copied log record from the receiver queue."""
        record = self._native.poll_record()
        if record is None:
            return None
        return LogRecord._from_native(record)


def set_log_callback(
    *,
    max_queued_records: int = 1024,
    consume: bool = False,
) -> LogReceiver:
    """Install a process-global bounded log-record receiver."""
    return LogReceiver._from_native(
        _native.set_log_callback(max_queued_records, consume)
    )


def clear_log_callback() -> None:
    """Clear the process-global native log callback."""
    _native.clear_log_callback()


def set_async_log_severity_mask(mask: LogSeverityMask) -> None:
    """Configure severities that may dispatch asynchronously."""
    _native.set_async_log_severity_mask(int(mask))


__all__ = [
    "LogEvent",
    "LogReceiver",
    "LogRecord",
    "LogSeverity",
    "LogSeverityMask",
    "clear_log_callback",
    "set_async_log_severity_mask",
    "set_log_callback",
]
